edge-modules/load-gen/Program.cs

```diff
@@ -30,42 +32,45 @@ static async Task Main()
-                        () => GenerateMessageAsync(moduleClient, batchId));
-
-                    // setup the twin update timer
-                    timers.Add(
-                        Settings.Current.TwinUpdateFrequency,
-                        Settings.Current.JitterFactor,
-                        () => GenerateTwinUpdateAsync(moduleClient, batchId));
-
-                    timers.Start();
-                    (CancellationTokenSource cts, ManualResetEventSlim completed, Option<object> handler) = ShutdownHandler.Init(TimeSpan.FromSeconds(5), Logger);
-                    Logger.LogInformation("Load gen running.");
-
-                    await cts.Token.WhenCanceled();
-                    Logger.LogInformation("Stopping timers.");
-                    timers.Stop();
-                    Logger.LogInformation("Closing connection to Edge Hub.");
-                    await moduleClient.CloseAsync();
-
-                    completed.Set();
-                    handler.ForEach(h => GC.KeepAlive(h));
-                    Logger.LogInformation("Load Gen complete. Exiting.");
-                }
+                Guid batchId = Guid.NewGuid();
+                Logger.LogInformation($"Batch Id={batchId}");
+
+                // setup the message PeriodicTask
+                messageTask = Option.Some(new PeriodicTask(
+                    () => GenerateMessageAsync(moduleClient, batchId),
+                    Settings.Current.MessageFrequency,
+                    TimeSpan.FromSeconds(Settings.Current.StartDelay),
+                    Logger,
+                    "Generate message"));
+
+                // setup the twin update PeriodicTask
+                twinUpdateTask = Option.Some(new PeriodicTask(
+                    () => GenerateTwinUpdateAsync(moduleClient, batchId),
+                    Settings.Current.MessageFrequency,
+                    TimeSpan.FromSeconds(Settings.Current.StartDelay),
+                    Logger,
+                    "Generate twin update"));
+
+                (CancellationTokenSource cts, ManualResetEventSlim completed, Option<object> handler) = ShutdownHandler.Init(TimeSpan.FromSeconds(5), Logger);
+                Logger.LogInformation("Load gen running.");
+
+                await cts.Token.WhenCanceled();
+                Logger.LogInformation("Closing connection to Edge Hub.");
+                await moduleClient.CloseAsync();
+
+                completed.Set();
+                handler.ForEach(h => GC.KeepAlive(h));
+                Logger.LogInformation("Load Gen complete. Exiting.");
             }
             catch (Exception ex)
             {
                 Logger.LogError($"Error occurred during load gen.\r\n{ex}");
             }
+            finally
+            {
+                messageTask.ForEach(x => x.Dispose());
+                twinUpdateTask.ForEach(x => x.Dispose());
+            }
         }
 
         static async Task GenerateMessageAsync(ModuleClient client, Guid batchId)
```

Comment:
This function no longer needs Interlocked if it is a periodic task. It can also use messageIdCounter directly.